### PR TITLE
Overhauling Coin Selection To Enable Multichange Selection At 5kb Output Edge Cases

### PR DESCRIPTION
**Pull request by NicholasMaselli**
This PR overhauls coin selection algorithm and change selection.

- Before we were only checking the ADA inputs after the balance was checked for inputs. But this could fail as change outputs with tokens could exist that we need to account for after selecting ada.
- As a result of the above, this new algorithm which follows similarly to the Nami wallet implementation with some improvements (https://github.com/berry-pool/nami/blob/b5e7280dd3204360160959dfa1d548eec97918a0/src/lib/coinSelection.js#L522).

The algorithm works as follows:
1. Perform the same logic for asset selection with bundles (but reverse values for mint and burn multiassets)
2. Immediately perform an ada input selection and change selection.
3. Ensure the exact input - output required change is greater than the minUTXO change input outputs.
4. If not perform another ada selection by adding the difference in input - output change from the require change. 
5. Loop step 3 and 4 until either the change is correct or until we run out of UTXOs or until we cannot create a balance greater than the ada amount. The Nami wallet uses an if statement here instead of a loop which can miss corner cases with high NFT counts and low ada amounts.

This PR also adds several parameters to improve the coin selection algorithms:
- changeAddress: This parameter automatically sets the output change address so we do not need to do this outside of the coin selection functions.
- mint: This optional token bundle builder is included to allow the coin selections to support both burning and minting. Using outputPurpose we were not able to support burning because we could never guarentee that a UTXO would include the burned token because Cardano outputs cannot have negative numbers.
- feeBuffer: This parameter adds a buffer for fees to ensure we have enough ada in our selected inputs to satisfy Cardano fees. 

In addition, the CIP2 tests have been updated and are working on tests. This is also working in live production tests =D

### CODE REVIEW

File: CardanoSharp.Wallet/CardanoSharp.Wallet.csproj [hunk 1]
```diff
@@ -4,7 +4,7 @@
     <TargetFrameworks>netstandard2.0;netstandard2.1</TargetFrameworks>
     <LangVersion>9.0</LangVersion>
     <GeneratePackageOnBuild>true</GeneratePackageOnBuild>
-    <Version>3.0.0</Version>
+    <Version>3.0.1</Version>
```
Comment (safestak-keith) [inline CardanoSharp.Wallet/CardanoSharp.Wallet.csproj:7]: There seems to be breaking changes to the main SDK surface (i.e. IChangeCreationStrategy, ICoinSelectionService, TransactionOutputExtensions etc.) which should be a major version bump?

Comment (NicholasMaselli) [reply CardanoSharp.Wallet/CardanoSharp.Wallet.csproj:7]: This PR was really supposed to be out with 3.0.0 but was split for the purposes of not having an 8000 line PR

Comment (nothingalike) [reply CardanoSharp.Wallet/CardanoSharp.Wallet.csproj:7]: we can bump to `4.0.0` no issue with that

Comment (safestak-keith) [reply CardanoSharp.Wallet/CardanoSharp.Wallet.csproj:7]: Yeah I would suggest `4.0.0` I think there are quite a few consumers using these methods (myself included)

File: CardanoSharp.Wallet/CIPs/CIP2/Extensions/TransactionOutputExtensions.cs [hunk 2]
```diff
@@ -42,17 +42,45 @@ public static Balance AggregateAssets(this IEnumerable<TransactionOutput> transa
...
                         }
 
-                        if(o.OutputPurpose == OutputPurpose.Burn) {
+                        nativeAsset.Quantity = nativeAsset.Quantity + na.Value;
+                    }
+                }
+            }
+
+            // remove / add assets from balance based on mint / burn token bundle
+            if (mint is not null) {
```
Comment (safestak-keith) [inline CardanoSharp.Wallet/CIPs/CIP2/Extensions/TransactionOutputExtensions.cs:51]: A nit but I think an inverse condition `if (mint is null) return balance;` would help with the readability (less indents)

File: CardanoSharp.Wallet/CIPs/CIP2/CoinSelectionUtility.cs [hunk 3]
```diff
@@ -10,18 +10,18 @@ namespace CardanoSharp.Wallet.CIPs.CIP2
...
+        public static CoinSelection UseRandomImprove(this TransactionBodyBuilder tbb, List<Utxo> utxos, string changeAddress, ITokenBundleBuilder mint = null, int limit = 20, ulong feeBuffer = 0)
         {
-            var cs = new CoinSelectionService(new RandomImproveStrategy(), new SingleTokenBundleStrategy());
+            var cs = new CoinSelectionService(new RandomImproveStrategy(), new BasicChangeSelectionStrategy());
             var tb = tbb.Build();
-            return cs.GetCoinSelection(tb.TransactionOutputs.ToList(), utxos);
+            return cs.GetCoinSelection(tb.TransactionOutputs.ToList(), utxos, changeAddress, mint, limit, feeBuffer);
         }
     }
-}
```
Comment (nothingalike) [inline CardanoSharp.Wallet/CIPs/CIP2/CoinSelectionUtility.cs:27]: the purpose of this class may not be clear. its suppose to help abstract away the configuring of different input and change strategies. so we dont need to really update unless parameters are different. really we just add to this.

Comment (NicholasMaselli) [reply CardanoSharp.Wallet/CIPs/CIP2/CoinSelectionUtility.cs:27]: My reason for updating is the SingleTokenBundle strategy will creating failing transactions for wallets with large amounts of NFTs. I haven't had any issues with the new strategy so wanted to make sure anyone using this function doesn't have failing tx's unexpectedly

File: CardanoSharp.Wallet/CardanoSharp.Wallet.csproj (hunk 1 above)
Comment (NicholasMaselli) [reply CardanoSharp.Wallet/CardanoSharp.Wallet.csproj]: 👍

**Review by nothingalike — APPROVED** [review]
LGTM

**Review by Daniel-SchaeferJ — APPROVED** [review]
Nice!

**Review by safestak-keith — APPROVED** [review]
LGTM

**Pull request merged**